libarchive/archive_write_set_format_7zip.c

```diff
@@ -420,16 +455,51 @@ _7z_options(struct archive_write *a, const char *key, const char *value)
 		return (ARCHIVE_OK);
 	}
 	if (strcmp(key, "compression-level") == 0) {
-		if (value == NULL ||
-		    !(value[0] >= '0' && value[0] <= '9') ||
-		    value[1] != '\0') {
-			archive_set_error(&(a->archive),
-			    ARCHIVE_ERRNO_MISC,
-			    "Illegal value `%s'",
-			    value);
+		if (value == NULL || *value == '\0') {
+			archive_set_error(&(a->archive), ARCHIVE_ERRNO_MISC,
+				"Invalid compression-level option value `%s'", value);
+			return (ARCHIVE_FAILED);
+		}
+
+		errno = 0;
+		char *end = NULL;
+		int lvl = (int)strtol(value, &end, 10);
```

Comment:
I just noticed the proposed strtol changes in #2110 - should I update this PR to match?

Reply:
@mostynb @kientzle yes, please

Reply:
Updated.